From 9e2be65f3034571e9aae5e28f20411898cb5d031 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Mon, 27 May 2024 18:42:55 +0200
Subject: [PATCH 0828/1695] drivers: rstctrl: support of IWDG timeout event to
 reset the application

Add support of IWDG event configuration to support reset of application
when the IWDG timeout occur.

This behavior will be select in board device tree for example:

&iwdg2 {
	resets = <&rcc IWDG2_SYS_R>;
};
=> system reset on IWDG2 timeout

OR

&iwdg2 {
	resets = <&rcc IWDG2_KER_R>;
};
=> only reset of CPU1

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 97665da58c7a6bb35f6bfd5db1a1460a51a557d6 drivers: rstctrl: support of IWDG timeout event to reset the application]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: I1fdfc3d7e4eae165359239f5e7fd2dd46dbb2dc2
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/382940
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Gabriel FERNANDEZ <gabriel.fernandez@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/rstctrl/stm32mp25_rstctrl.c          | 4 ++++
 core/include/dt-bindings/reset/stm32mp25-resets.h | 6 +++++-
 2 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/core/drivers/rstctrl/stm32mp25_rstctrl.c b/core/drivers/rstctrl/stm32mp25_rstctrl.c
index 86420174a..84092b555 100644
--- a/core/drivers/rstctrl/stm32mp25_rstctrl.c
+++ b/core/drivers/rstctrl/stm32mp25_rstctrl.c
@@ -222,6 +222,10 @@ const struct stm32_reset_cfg *stm32mp25_reset_cfg[STM32MP25_LAST_RESET] = {
 
 	RST_SETR(IWDG2_KER_R,	RCC_IWDGC1CFGSETR,	18),
 	RST_SETR(IWDG4_KER_R,	RCC_IWDGC2CFGSETR,	18),
+	RST_SETR(IWDG1_SYS_R,	RCC_IWDGC1CFGSETR,	0),
+	RST_SETR(IWDG2_SYS_R,	RCC_IWDGC1CFGSETR,	2),
+	RST_SETR(IWDG3_SYS_R,	RCC_IWDGC2CFGSETR,	0),
+	RST_SETR(IWDG4_SYS_R,	RCC_IWDGC2CFGSETR,	2),
 
 	RST_INV(C2_HOLDBOOT_R,	RCC_CPUBOOTCR,		0),
 	RST_INV(C1_HOLDBOOT_R,	RCC_CPUBOOTCR,		1),
diff --git a/core/include/dt-bindings/reset/stm32mp25-resets.h b/core/include/dt-bindings/reset/stm32mp25-resets.h
index 7868f1425..be3ee56ee 100644
--- a/core/include/dt-bindings/reset/stm32mp25-resets.h
+++ b/core/include/dt-bindings/reset/stm32mp25-resets.h
@@ -151,8 +151,12 @@
 #define DDRPHYCAPB_R	141
 #define DDRCFG_R	142
 #define DDR_R		143
+#define IWDG1_SYS_R	144
+#define IWDG2_SYS_R	145
+#define IWDG3_SYS_R	146
+#define IWDG4_SYS_R	147
 
-#define STM32MP25_LAST_RESET	144
+#define STM32MP25_LAST_RESET	148
 
 #define RST_SCMI_C1_R		0
 #define RST_SCMI_C2_R		1
-- 
2.47.2

